### ISSUE TITLE
Breaks on Node 16

### ISSUE THREAD

**Issue by hexeberlin** [sinonjs/fake-timers#394]
Currently, the tests are failing in node 16. Figure out why, fix it, and adjust GitHub actions workflow to use node 16 instead of 10.

**Comment by fatso83** [comment]
@mroderick Thoughts on this? Just came across this. Might tangentially touch work currently done by @itayperry .

**Comment by benjamingr** [comment]
Node.js 16 has global `performance` so it makes sense it's behaving differently than older versions here.

**Comment by fatso83** [comment]
What does that mean? Does it means it is not present on `global` anymore? We detect its presence before installing the timers on Node 16 as well, so not really getting it.

**Comment by mroderick** [comment]
In node 16

```
> typeof Performance
'undefined'
> typeof global.performance
'object'
> typeof performance
'object'
```

**Comment by mroderick** [comment]
> @mroderick Thoughts on this (ref [0b1834a](https://github.com/sinonjs/fake-timers/commit/0b1834a70ca228ab6cc5e9df5f24d072c5d11ace))?

@hexeberlin and I made all the sinon repositories use GitHub Actions in the same day. In the parent commit b3557efac80a77780eb2349f8313d074ef7d4149, you can see this. Unfortunately, we couldn't run the `fake-timers` tests in Node 16, so we settled for 10-14 and filed a bug to fix it later.

**Comment by itayperry** [comment]
This returns `true` today when running the test that @fatso83 has mentioned:
https://github.com/sinonjs/fake-timers/blob/93f3f626593f646535d1045fcb517cb6a107253f/src/fake-timers-src.js#L154-L156

> typeof Performance
'undefined'

Was it available on node 14 and deleted on node 16...? (I don't really get it). 
@benjamingr

**Comment by fatso83** [comment]
@itayperry The stuff you did in #400 seemed to touch upon this. Want to merge that before we look more into this?

**Comment by synaestheory** [comment]
> ~This returns `true` today when running the test that @fatso83 has mentioned:~ I guess this is what's causing the problem - it is expected to be `true` in order to stub `performance`:
> 
> https://github.com/sinonjs/fake-timers/blob/93f3f626593f646535d1045fcb517cb6a107253f/src/fake-timers-src.js#L154-L156
> 
> > typeof Performance
> > 'undefined'
> 
> ~Was it available on node 14 and deleted on node 16...?~ ~(I don't really get it).~ @benjamingr

The major difference between node 14 and node 16 appears to be the introduction of [global.performance](https://nodejs.org/docs/latest-v16.x/api/globals.html#performance) in v16.

In v14, [performancePresent](https://github.com/sinonjs/fake-timers/blob/93f3f626593f646535d1045fcb517cb6a107253f/src/fake-timers-src.js#L152) resolves to `undefined` but in v16 it resolves to `true`. 

One other delta that I noticed while stepping through the code was that `hijackMethod` is also called for `global.performance` in v16, where it is not in v14. I'm not suggesting that this is correct/incorrect behavior, just noting it as a delta.

**Comment by itayperry** [comment]
@fatso83 
> Want to merge that before we look more into this?
 
Yes, of course! I'd be happy to do it :) I'll speed it up.

**Comment by fatso83** [comment]
@itayperry Any chances of you looking into this? Seems you already did all the work 😄

**Comment by itayperry** [comment]
@fatso83 
On it :) I'll probably ask for some advice, I remember there were issues with `performance.mark`.
I'm looking into this 👾

**Comment by itayperry** [comment]
Hi, I don't know if I'm doing this right.. but I'm using `nvm`, and when I try `nvm use 14` or `nvm use 15` everything works - when I try `nvm use 16`  I don't get 3 failed tests but **67** failed tests.

`npm run test-node`:

![image](https://user-images.githubusercontent.com/37377389/149638530-12d9eabb-4f70-4f24-91c5-a59286fd6191.png)

I guess I'm doing something wrong.

**Comment by itayperry** [comment]
About these 2 tests: 
a) https://github.com/sinonjs/fake-timers/blob/1b1ac4a781c57498b8aa9015c6448bd02e54830c/test/fake-timers-test.js#L3758
b) https://github.com/sinonjs/fake-timers/blob/1b1ac4a781c57498b8aa9015c6448bd02e54830c/test/fake-timers-test.js#L3778

They both use `Performance` (to be more accurate `Performance.prototype`) which does not exist in `Node 16`+.
I did a small research and found this class called: `InternalPerformance` in Node's source code:
```js
InternalPerformance.prototype.constructor = Performance.prototype.constructor;
ObjectSetPrototypeOf(InternalPerformance.prototype, Performance.prototype);
```
https://github.com/nodejs/node/blob/e2e2bc83c01b19a6601a3018d1790fbd2d4fa7de/lib/internal/perf/performance.js#L103-L104

```js
performance: new InternalPerformance()
```
https://github.com/nodejs/node/blob/e2e2bc83c01b19a6601a3018d1790fbd2d4fa7de/lib/perf_hooks.js#L36

 but I'm afraid I can't use it / access it.. it would be nice if we could use `InternalPerformance.prototype` instead of `Performance.prototype`. 
 
 What do you think we should do with these specific tests? Can we just skip them when in `Node` env?

**Comment by msluther** [comment]
I made a stupid workaround that _happens_ to work in my particular use case if anyone else is hitting this issue and wants to give it a try. I basically just made my own global fake `Performance` class and proxied to `performance`. It's dirty and hacky but it allowed my tests to continue working and I'm not blocked on upgrading to node@16.


I've got this in my unit test setup that gets pulled in prior to anyone calling `useFakeTimers`

```js
global.Performance = global.Performance || class Perf {
  clearMarks() { return global.performance.clearMarks(); }
  clearMeasures() { return global.performance.clearMeasures(); }
  clearResourceTimings() { return global.performance.clearResourceTimings(); }
  getEntries() { return global.performance.getEntries(); }
  getEntriesByName() { return global.performance.getEntriesByName(); }
  getEntriesByType() { return global.performance.getEntriesByType(); }
  mark() { return global.performance.mark(); }
  measure() { return global.performance.measure(); }
  now() { return global.performance.now(); }
  setResourceTimingBufferSize() { return global.performance.setResourceTimingBufferSize(); }
  toJSON() { return global.performance.toJSON(); }
};
```

**Comment by benjamingr** [comment]
Hey sorry this has been open for a while and I need this to work in my project so I'm going to go ahead and fix it - let me know if you want to sync on other stuff to work on.

**Comment by itayperry** [comment]
@benjamingr 
It would take me much longer than it took you to resolve it, so I get it. It's ok 🌼 
Plus, I didn't know if it were right to skip these 2 tests I was writing about..

Anyway, thank you for your help!!

**Comment by benjamingr** [comment]
Can be closed :)

**Comment by msluther** [comment]
It looks like this will need to be bumped over in [`nise`](https://github.com/sinonjs/nise) as well as I am still hitting this issue using fake timers via sinon. I'll open an issue there.

```sh-output
$ npm ls @sinonjs/fake-timers
@configstore/ui-components@0.2.0 /Users/mluther1/ghec/gdcorp-uxp/configstore-ui-components
└─┬ sinon@13.0.0
  ├── @sinonjs/fake-timers@9.0.0
  └─┬ nise@5.1.0
    └── @sinonjs/fake-timers@7.1.2
```

**Comment by msluther** [comment]
Hrm. So, the fix works but makes the assumption that the functions on `global.Performance.prototype` are a superset of `global.performance.*` functions. This is currently not the case when using `node@16` and `jsdom`. [`jsdom` is injecting](https://github.com/jsdom/jsdom/blob/master/lib/jsdom/browser/Window.js#L340) a `global.Performance` object based on the minimal W3C spec for [`hr-time`](https://w3c.github.io/hr-time/#sec-performance). 

This isn't necessarily a bug with either library, but feels like a gap between the two. 

_However_, `@sinon/fake-timers` ends up replacing a good `global.performance.mark` with `undefined` in this situation. Should the [code here](https://github.com/sinonjs/fake-timers/blob/master/src/fake-timers-src.js#L1671-L1687) use both prototypes to make sure this is the case or maybe prefer the `global.performance` over `global.Performance`?

The work around that I [posted above](https://github.com/sinonjs/fake-timers/issues/394#issuecomment-1021665619) still works in this case as long as its injected prior to sinon & jsdom, so I'm not blocked by any means, but just requires making sure that happens in my tests everywhere.

**Comment by benjamingr** [comment]
@msluther to understand your use case better - you are using a library (e.g. for testing) that sets your global object to a JSDom window?

If so - I think it's probably reasonable to account for then yes - can you open a separate issue?

**Comment by msluther** [comment]
Yep, thats correct its basically this in my test setup: `require('global-jsdom/register')` and that ends up with the minimal version of global.Performance type, but leaves `global.performance`.  I'll open another issue.  Thanks
